File: Python3/symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def check(root_l, root_r):
            if not root_l and not root_r:
                return True
            elif not root_l or not root_r:
                return False
            else:
                if root_l.val == root_r.val:
                    if check(root_l.left, root_r.right):
                        if check(root_l.right, root_r.left):
                            return True
                        else:
                            return False
                    else:
                        return False
                else:
                    return False
        
        ans = check(root.left, root.right)
        return ans
        

File: C++/symmetric_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        bool ans = check(root->left, root->right);
        return ans;
    }
private:
    bool check(TreeNode* root_r, TreeNode* root_l){
        if (!root_r && !root_l) {
            return true;
        } else if (!root_l || !root_r) {
            return false;
        } else {
            if (root_r->val == root_l->val) {
                if (check(root_l->left, root_r->right)){
                    if (check(root_l->right, root_r->left)) {
                        return true;
                    } else {
                        return false;
                    }
                } else {
                    return false;
                }
            } else {
                return false;
            }
        }
    }
};
